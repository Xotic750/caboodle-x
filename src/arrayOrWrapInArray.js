/**
 * @file Utility that assures a returned value of type Array.
 * @copyright Copyright (c) 2018-present, Graham Fairweather
 * @module arrayOrWrapInArray
 */

/**
 * Takes a single, optional argument and then performs the following logic:
 * If an argument is supplied and is an array then it will be returned unchanged,
 * if an argument is supplied and is not an array then it will be wrapped in a new array and returned,
 * if no argument is supplied then a new empty array is returned.
 *
 * @param {Array|*} [args] - A single optional identity argument.
 * @returns {Array} An array as described the the description logic.
 */
export default function arrayOrWrapInArray(...args) {
  if (args.length < 1) {
    return [];
  }

  const [identity] = args;

  return Array.isArray(identity) ? identity : [identity];
}
