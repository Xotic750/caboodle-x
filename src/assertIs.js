/**
 * @file Utility that needs description.
 * @copyright Copyright (c) 2018-present, Graham Fairweather
 * @module assertIs
 */

import isFunctionType from 'src/isFunctionType';

export default function assertIs(...args) {
  const [predicate, defaultMessage] = args;

  if (!isFunctionType(predicate)) {
    throw new TypeError('Predicate must be a function.');
  }

  const defMsg = args.length > 1 ? `${defaultMessage}` : 'Not a valid value.';

  return function assertIsBound(...boundArgs) {
    const [value, customMessage] = boundArgs;

    if (!predicate.call(this, value)) {
      const message = boundArgs.length > 1 ? `${customMessage}` : defMsg;

      throw new TypeError(message);
    }

    return value;
  };
}
