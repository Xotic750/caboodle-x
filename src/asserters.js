import {
  assertIs,
  isFunctionType,
  isObjectLikeNotArray,
  isPropertyKey,
} from 'src/validators';

export const assertIsObjectLikeNotArray = assertIs(isObjectLikeNotArray, 'Not a valid object.');

export const assertIsPropertyKey = assertIs(isPropertyKey, 'Not a valid property key.');

export const assertIsFunctionType = assertIs(isFunctionType, 'Not a function.');
