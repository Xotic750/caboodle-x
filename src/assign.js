/**
 * @file Utility that needs description.
 * @copyright Copyright (c) 2018-present, Graham Fairweather
 * @module assign
 */

import isFunctionType from './isFunctionType';
import isNil from './isNil';
import toObject from './toObject';
import stubArray from './stubArray';
import _concat from './.internal/_concat';
import _filter from './.internal/_filter';
import _reduce from './.internal/_reduce';
import _keys from './.internal/_keys';
import _propertyIsEnumerable from './.internal/_propertyIsEnumerable';
import _Object from './.internal/_Object';

/* istanbul ignore next */
const getOwnPropertySymbols = isFunctionType(_Object.getOwnPropertySymbols) ? _Object.getOwnPropertySymbols : stubArray;

const getOwnEnumerablePropertySymbols = function _getOwnEnumerablePropertySymbols(target) {
  const isEnumerable = function _isEnumerable(symbol) {
    return _propertyIsEnumerable(target, symbol);
  };

  return _filter(getOwnPropertySymbols(target), isEnumerable);
};

const reducer = function _reducer(tgt, source) {
  if (isNil(source)) {
    return tgt;
  }

  const object = _Object(source);
  const assigner = function _assigner(tar, key) {
    tar[key] = object[key];

    return tar;
  };

  return _reduce(_concat(_keys(object), getOwnEnumerablePropertySymbols(object)), assigner, tgt);
};

/**
 * This method is used to copy the values of all enumerable own properties from
 * one or more source objects to a target object. It will return the target object.
 *
 * @param {*} target - The target object.
 * @param {Array} rest - The rest of the arguments array.
 * @param {...*} [rest.source] - The source object(s).
 * @throws {TypeError} If target is null or undefined.
 * @returns {object} The target object.
 */

export default function assign(target, ...rest) {
  return _reduce(rest, reducer, toObject(target));
}
