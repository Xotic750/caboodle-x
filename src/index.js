/**
 * @file Re-exporter.
 * @copyright Copyright (c) 2018-present, Graham Fairweather
 * @module index
 */

export {default as arrayOrWrapInArray} from 'src/arrayOrWrapInArray';
export {default as assertIsFunctionType} from 'src/assertIsFunctionType';
export {default as assertIsObject} from 'src/assertIsObject';
export {default as assertIs} from 'src/assertIs';
export {default as callFunctionOrIdentity} from 'src/callFunctionOrIdentity';
export {default as capitalizeFirst} from 'src/capitalizeFirst';
export {default as capitalize} from 'src/capitalize';
export {default as clamp} from 'src/clamp';
export {default as constant} from 'src/constant';
export {default as defaultToOneOf} from 'src/defaultToOneOf';
export {default as defineValidatorProperties} from 'src/defineValidatorProperties';
export {default as defineValidatorProperty} from 'src/defineValidatorProperty';
export {default as delayPromise} from 'src/delayPromise';
export {default as es5IdentifierRx} from 'src/es5IdentifierRx';
export {default as es5ReservedNamesRx} from 'src/es5ReservedNamesRx';
export {default as es6IdentifierRx} from 'src/es6IdentifierRx';
export {default as es6ReservedNamesRx} from 'src/es6ReservedNamesRx';
export {default as identity} from 'src/identity';
export {default as includesNonWhitespace} from 'src/includesNonWhitespace';
export {default as includesWhitespace} from 'src/includesWhitespace';
export {default as isArrayLikeObject} from 'src/isArrayLikeObject';
export {default as isArrayLike} from 'src/isArrayLike';
export {default as isBinary} from 'src/isBinary';
export {default as isBooleanType} from 'src/isBooleanType';
export {default as isCamecaseIdentifier} from 'src/isCamecaseIdentifier';
export {default as isConstantIdentifier} from 'src/isConstantIdentifier';
export {default as isCountingNumber} from 'src/isCountingNumber';
export {default as isDateObject} from 'is-date-object';
export {default as isFunctionType} from 'src/isFunctionType';
export {default as isInvalidHexLiteral} from 'src/isInvalidHexLiteral';
export {default as isNil} from 'src/isNil';
export {default as isNull} from 'src/isNull';
export {default as isNumberType} from 'src/isNumberType';
export {default as isObjectLikeNotArray} from 'src/isObjectLikeNotArray';
export {default as isObjectLike} from 'src/isObjectLike';
export {default as isObjectType} from 'src/isObjectType';
export {default as isObject} from 'src/isObject';
export {default as isOctal} from 'src/isOctal';
export {default as isOneOf} from 'src/isOneOf';
export {default as isPopulatedString} from 'src/isPopulatedString';
export {default as isPrimitive} from 'is-primitive';
export {default as isReservedEs5Identifier} from 'src/isReservedEs5Identifier';
export {default as isReservedEs6Identifier} from 'src/isReservedEs6Identifier';
export {default as isReservedIdentifier} from 'src/isReservedIdentifier';
export {default as isSearchIndex} from 'src/isSearchIndex';
export {default as isSpaced} from 'src/isSpaced';
export {default as isStringTypeOrInteger} from 'src/isStringTypeOrInteger';
export {default as isStringTypeOrNumberType} from 'src/isStringTypeOrNumberType';
export {default as isStringType} from 'src/isStringType';
export {default as isSymbolType} from 'src/isSymbolType';
export {default as isSymbol} from 'is-symbol';
export {default as isUndefined} from 'src/isUndefined';
export {default as isUnderscored} from 'src/isUnderscored';
export {default as isUniq} from 'src/isUniq';
export {default as isValidEs5Identifier} from 'src/isValidEs5Identifier';
export {default as isValidEs6Identifier} from 'src/isValidEs6Identifier';
export {default as isValidHtmlId} from 'src/isValidHtmlId';
export {default as isValidIdentifier} from 'src/isValidIdentifier';
export {default as isWholeNumber} from 'src/isWholeNumber';
export {default as kebabJoin} from 'src/kebabJoin';
export {default as nilifyIs} from 'src/nilifyIs';
export {default as noop} from 'src/noop';
export {default as parseDecimal} from 'src/parseDecimal';
export {default as sameValueZero} from 'src/sameValueZero';
export {default as stubArray} from 'src/stubArray';
export {default as stubFalse} from 'src/stubFalse';
export {default as stubObject} from 'src/stubObject';
export {default as stubString} from 'src/stubString';
export {default as stubTrue} from 'src/stubTrue';
export {default as toStringTag} from 'src/toStringTag';
export {default as uniq} from 'src/uniq';
