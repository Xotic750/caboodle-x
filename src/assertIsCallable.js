/**
 * @file Utility that needs description.
 * @copyright Copyright (c) 2018-present, Graham Fairweather
 * @module assertIsCallable
 */

import ES from 'es-abstract';
import assertIs from 'src/assertIs';

export default assertIs(ES.IsCallable, 'Not callable.');
