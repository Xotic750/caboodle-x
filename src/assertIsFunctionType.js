/**
 * @file Utility that needs description.
 * @copyright Copyright (c) 2018-present, Graham Fairweather
 * @module assertIsFunctionType
 */

import isFunctionType from 'src/isFunctionType';
import assertIs from 'src/assertIs';

export default assertIs(isFunctionType, 'Not a function.');
