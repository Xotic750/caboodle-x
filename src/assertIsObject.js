/**
 * @file Utility that needs description.
 * @copyright Copyright (c) 2018-present, Graham Fairweather
 * @module assertIsObject
 */

import assertIs from 'src/assertIs';
import isObject from 'src/isObject';

export default assertIs(isObject, 'Not an object.');
