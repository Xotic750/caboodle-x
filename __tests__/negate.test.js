import {negate} from 'src/caboodle-x';

const stubTrue = () => true;
const isEven = (n) => n % 2 === 0;

describe('negate', () => {
  it('is a function', () => {
    expect(typeof negate).toBe('function');
  });

  it('should throw when array is null or undefined', () => {
    expect(() => {
      negate();
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      negate(undefined);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      negate(null);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should create a function that negates the result of `func`', () => {
    const negateFn = negate(isEven);

    expect(negateFn(1)).toBe(true);
    expect(negateFn(2)).toBe(false);
  });

  it('should create a function that accepts multiple arguments', () => {
    let argCount;
    const count = 5;
    const negateFn = negate(function() {
      argCount = arguments.length;
    });
    const expected = Array(count)
      .fill()
      .map(stubTrue);

    const actual = Array(count)
      .fill()
      .map((unused, index) => {
        switch (index) {
          case 0:
            negateFn();
            break;

          case 1:
            negateFn(1);
            break;

          case 2:
            negateFn(1, 2);
            break;

          case 3:
            negateFn(1, 2, 3);
            break;

          case 4:
          default:
            negateFn(1, 2, 3, 4);
        }

        return argCount === index;
      });

    expect(actual).toStrictEqual(expected);
  });
});
