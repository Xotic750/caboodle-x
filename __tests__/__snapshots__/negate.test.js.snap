// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`negate should throw when array is null or undefined 1`] = `"Not a function"`;

exports[`negate should throw when array is null or undefined 2`] = `"Not a function"`;

exports[`negate should throw when array is null or undefined 3`] = `"Not a function"`;
